### PR TITLE
feat(storage): support compaction group split and merge

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previously, we implemented the concept of compaction groups for hummock to achieve write/configuration isolation and to create new compaction groups via the table-level move function.

Previously, we implemented the concept of compaction groups for hummock to achieve write/configuration isolation and to create new compaction groups via the table-level move function.

Since the sst key range generated by the move function is not contiguous, this makes it difficult to re-combine the resulting independent compaction groups (currently not supported). Because of this concern, we set the move threshold very high, fearing that the resulting compaciton group will not be able to be merged again, thus creating a large number of compaction groups.

This concern is contrary to our expectation that in scenarios with high write throughput (backfill, etc.), we would like to split the compaction to achieve parallel compaction, increase the overall compaction throughput, and make full use of the compactor node's scale capability.

Therefore, we redesigned hummock's scalability by implementing split / merge instead of move, in order to obtain the following capabilities.

1. adopt more lenient split conditions to cope with high write/burst write traffic scenarios, to extend compaction capability horizontally.
2. auto-merge compaction groups after write throughput is restored.
3. refine the granularity of Split from table to key_range, to split huge compaction groups at a finer granularity
4. Better O&M tools to manually split and merge compaction groups.

For the current PR, we expect to implement table-level Split/Merge capability (phase 1), and plan to further implement Key-level Split in the subsequent PRs, and make CN not aware of the concept of group_id. Here are the changes involved in this PR for review.

1. add the concept of estimated_size to SstableInfo for more accurate Compaction policy calculation.
2. based on table_id level Split function, compatible with the original Split interface.
- Fix key_range/table_id/estimated size after sst split.
- Use (table_id, vnode = 0, epoch = 0) as split key for splitting.
3. merge compaction group
4. schedule capability for automatic split/merge.
5. compactor iterator support key_range filter

I'll summarise the changes to the file, for your code reivew:
1. src/meta/src/hummock/manager/commit_epoch.rs adapt the logic of commit_epoch so that all NewTableFragmentInfo creates a new compaction group
2. src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs introduce helper functions that facilitate splitting, refactor the split interface based on the new implementation, and provide a simple auto split/merge strategy. (The auto split strategy is based on the previous design).
3. src/meta/src/hummock/manager/timer_task.rs  periodically call the try function provided by schedule to implement auto split/merge.
4. src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs reimplemented split logic based on split_key.
5. src/storage/hummock_sdk/src/compaction_group/mod.rs provides a wrapper for the split_key based sst / level split implementation.
6. src/storage/hummock_sdk/src/version.rs removed cg2 / cg3 special assumptions.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/catalog/system_catalog/rw_catalog/rw_hummock_version.rs [hunk 1]
```diff
@@ -124,7 +125,7 @@ fn version_to_sstable_rows(version: HummockVersion) -> Vec<RwHummockSstable> {
                     key_range_left: key_range.left.to_vec(),
                     key_range_right: key_range.right.to_vec(),
                     right_exclusive: key_range.right_exclusive,
-                    file_size: sst.file_size as _,
+                    file_size: estimated_sst_file_size as _,
```
[the remark this answers is not in the document]
Comment (Li0k) [reply src/frontend/src/catalog/system_catalog/rw_catalog/rw_hummock_version.rs:128]: Thank you for bringing it to my attention, I'll be fixing each of these details as soon as I finish the prototype.

**Review by hzxa21 — COMMENTED** [review]
Almost finish the review. Let me post the comments I have so far first before I continue to add more.

**Comment by hzxa21** [comment]
Btw, I think the PR is larger than I expected. To reduce the risk of this PR, is it possible to further breaking it into small ones so that we can merge small PRs one by one.

**Comment by Li0k** [comment]
> Btw, I think the PR is larger than I expected. To reduce the risk of this PR, is it possible to further breaking it into small ones so that we can merge small PRs one by one.

Thanks, let's review the code framework first  before "ready to review"

**Pull request closed without merging**